From 39cd0c75cd4021f6de3812dfbcecf1e0a65cb310 Mon Sep 17 00:00:00 2001
From: Changqing Li <changqing.li@windriver.com>
Date: Wed, 21 May 2025 20:21:59 +0800
Subject: [PATCH] Add url encoded username/password entry

With the latest two commits in git, refer [1][2], git prompt will url
encode the user and host, anspass should align with it, otherwise, the
saved entry will not match the git prompt, this will cause
"Authentication failed", and make wrlinux setup report error:
fatal: Authentication failed for 'https://xxx'
The saved user or password seems invalid, remove it and try again?

Since these two commits related to CVE fix, some low version of git also
possbile include these commits. And it is possible that some users
include these commits and some users don't include these commits,
anspass need to be compatibel with both cases, so we will insert both
encoded and unencoded username/password entry

[1] https://github.com/git/git/commit/7725b8100ffbbff2750ee4d61a0fcc1f53a086e8
[2] https://github.com/git/git/commit/c903985bf7e772e2d08275c1a95c8a55ab011577

Upstream-Status: Submitted [https://github.com/Wind-River/anspass/pull/6]

Signed-off-by: Changqing Li <changqing.li@windriver.com>
---
 anspass-lib.c |  22 +++++
 anspass-lib.h |   4 +
 anspassd.c    | 258 ++++++++++++++++++++++++++++++++++++++++++++++----
 3 files changed, 265 insertions(+), 19 deletions(-)

diff --git a/anspass-lib.c b/anspass-lib.c
index cd0a01e..9398c04 100644
--- a/anspass-lib.c
+++ b/anspass-lib.c
@@ -32,6 +32,7 @@
 #include <sys/un.h>
 #include <termios.h>
 #include <unistd.h>
+#include <ctype.h>
 
 int is_env_set() {
 	if (getenv(ANSPASS_ENV) != NULL)
@@ -182,3 +183,24 @@ int get_secret(unsigned char *text, uint max, const int echo) {
 	tcsetattr( STDIN_FILENO, TCSANOW, &old);
 	return i;
 }
+
+#define URL_UNSAFE_CHARS " <>\"%{}|\\^`:?#[]@!$&'()*+,;="
+//This function use the same encode way as Git
+void str_percentencode(char *dst, const char *src, int flags)
+{
+	size_t i, len = strlen(src);
+
+	for (i = 0; i < len; i++) {
+		unsigned char ch = src[i];
+		if (ch <= 0x1F || ch >= 0x7F ||
+			(ch == '/' && (flags & ENCODE_SLASH)) ||
+			((flags & ENCODE_HOST_AND_PORT) ?
+			!isalnum(ch) && !strchr("-.:[]", ch) :
+			!!strchr(URL_UNSAFE_CHARS, ch))) {
+				sprintf(dst, "%%%02X", (unsigned char)ch);
+				dst += 3;
+			}
+		else
+			*dst++ = ch;
+	}
+}
diff --git a/anspass-lib.h b/anspass-lib.h
index 9750b37..75812bc 100644
--- a/anspass-lib.h
+++ b/anspass-lib.h
@@ -48,6 +48,9 @@
 #define SHUTDOWN                7
 #define ACK                     104
 
+#define ENCODE_SLASH 1
+#define ENCODE_HOST_AND_PORT 2
+
 struct anspass_packet {
 	int socket;
 	struct timeval *to;
@@ -82,4 +85,5 @@ unsigned char* string_to_hex(char * str);
 
 int get_secret(unsigned char* text, const uint max, const int echo);
 void print_env(const char* missing);
+void str_percentencode(char *dst, const char *src, int flags);
 #endif
diff --git a/anspassd.c b/anspassd.c
index 59e8044..7b1819b 100644
--- a/anspassd.c
+++ b/anspassd.c
@@ -31,8 +31,6 @@
 #include <sys/un.h>
 #include <sys/stat.h>
 
-
-
 int main(int argv, char *argc[]) {
 	int ret = -EINVAL;
 	int i = 0;
@@ -1213,6 +1211,80 @@ oom:
 	return;
 }
 
+static inline void get_percentencode_username_str(char *user, struct anspass_packet *in) {
+	char *token;
+	char *d = "/";
+
+	char *msg = (char*)calloc(1, strlen(in->msg) + 1);
+	if (!msg)
+		goto oom;
+	memcpy(msg, in->msg, strlen(in->msg) + 1);
+
+	token = strtok(msg, d);
+	sprintf(user, "%s'%s//", USERNAME_STR, token);
+
+	// Try to encode host in username prompt
+	char* encoded_str = (char*)calloc(1, 3*MAX_MSG_LENGTH + 1);
+	token = strtok(NULL, d);
+	if(token) {
+		str_percentencode(encoded_str, token, ENCODE_HOST_AND_PORT);
+		strcat(user, encoded_str);
+		memset(encoded_str, 0, 3*MAX_MSG_LENGTH + 1);
+	}
+	token = strtok(NULL, d);
+	while(token) {
+		strcat(user, "/");
+		str_percentencode(encoded_str, token, ENCODE_HOST_AND_PORT);
+		strcat(user, encoded_str);
+		memset(encoded_str, 0, 3*MAX_MSG_LENGTH + 1);
+		token = strtok(NULL, d);
+	}
+	strcat(user, "': ");
+	free(msg);
+	free(encoded_str);
+oom:
+	return;
+}
+
+static inline void get_percentencode_passwd_str(char *passwd, struct anspass_packet *in) {
+	char *token;
+	char *d = "/";
+
+	char *msg = (char*)calloc(1, strlen(in->msg) + 1);
+	if (!msg)
+		goto oom;
+	memcpy(msg, in->msg, strlen(in->msg) + 1);
+
+	token = strtok(msg, d);
+	sprintf(passwd, "%s'%s//", PASSWORD_STR, token);
+	char* encoded_str = (char*)calloc(1, 3*MAX_MSG_LENGTH + 1);
+	//Try to encode username in password prompt
+	str_percentencode(encoded_str, in->user, ENCODE_SLASH);
+	strcat(passwd, encoded_str);
+	memset(encoded_str, 0, 3*MAX_MSG_LENGTH + 1);
+	strcat(passwd, "@");
+	//Try to encode host in password prompt
+	token = strtok(NULL, d);
+	if(token) {
+		str_percentencode(encoded_str, token, ENCODE_HOST_AND_PORT);
+		strcat(passwd, encoded_str);
+		memset(encoded_str, 0, 3*MAX_MSG_LENGTH + 1);
+	}
+	token = strtok(NULL, d);
+	while(token) {
+		strcat(passwd, "/");
+		str_percentencode(encoded_str, token, ENCODE_HOST_AND_PORT);
+		strcat(passwd, encoded_str);
+		memset(encoded_str, 0, 3*MAX_MSG_LENGTH + 1);
+		token = strtok(NULL, d);
+	}
+	strcat(passwd, "': ");
+	free(msg);
+	free(encoded_str);
+oom:
+	return;
+}
+
 void process_request(struct anspass_packet *in) {
 
 	syslog(LOG_PID | LOG_DEBUG, "Received packet: \"%s\"\n", in->msg);
@@ -1293,6 +1365,8 @@ int del (struct anspass_packet *in){
 	struct db1_entry *e_user = NULL;
 	struct db1_entry *e_passwd = NULL;
 
+	struct db1_entry *e_encoded_user = NULL;
+	struct db1_entry *e_encoded_passwd = NULL;
 
 	char* passwd = (char*)calloc(1, get_passwd_len(in->msg, in->user));
 	if (!passwd)
@@ -1302,12 +1376,23 @@ int del (struct anspass_packet *in){
 	if (!user)
 		goto no_user_mem;
 
+	char* encoded_passwd = (char*)calloc(1, 3 * get_passwd_len(in->msg, in->user));
+	if (!encoded_passwd)
+		goto no_encoded_passwd_mem;
+	char* encoded_user = (char*)calloc(1, 3 * (sizeof(char) * (strlen(in->msg) +
+				strlen(USERNAME_STR) + 5)));
+	if (!encoded_user)
+		goto no_encoded_user_mem;
+
 	/* Username string is easy */
 	sprintf(user, "%s'%s': ", USERNAME_STR, in->msg);
 
+	get_percentencode_username_str(encoded_user, in);
 	/* Passwd has to insert username into the URI */
 	get_passwd_str(passwd, in);
 
+	get_percentencode_passwd_str(encoded_passwd, in);
+
 	ret = get_db1_entry(user, &e_user);
 	if (ret)
 		goto user_dne;
@@ -1316,6 +1401,19 @@ int del (struct anspass_packet *in){
 	if (ret)
 		goto passwd_dne;
 
+	if (strcmp(user, encoded_user) != 0) {
+		ret = get_db1_entry(encoded_user, &e_encoded_user);
+		if (ret) {
+			goto encoded_user_dne;
+		}
+	}
+	if (strcmp(passwd, encoded_passwd) != 0) {
+		ret = get_db1_entry(encoded_passwd, &e_encoded_passwd);
+		if (ret) {
+			goto encoded_passwd_dne;
+		}
+	}
+
 	ret = del_db2_cred(e_user);
 	if (ret)
 		goto del_db2_user_fail;
@@ -1334,8 +1432,42 @@ int del (struct anspass_packet *in){
 		goto del_passwd_fail;
 	e_passwd = NULL;
 
+	if (strcmp(user, encoded_user) != 0) {
+		ret = del_db2_cred(e_encoded_user);
+		if (ret)
+			goto del_db2_encoded_user_fail;
+		ret = del_db1_entry(e_encoded_user);
+		if (ret)
+			goto del_encoded_user_fail;
+	}
+	e_encoded_user = NULL;
+
+	if (strcmp(passwd, encoded_passwd) != 0) {
+		ret = del_db2_cred(e_encoded_passwd);
+		if (ret)
+			goto del_db2_encoded_passwd_fail;
+		ret = del_db1_entry(e_encoded_passwd);
+		if (ret)
+			goto del_encoded_passwd_fail;
+	}
+	encoded_passwd = NULL;
+
 	ret = 0;
 
+del_encoded_passwd_fail:
+del_db2_encoded_passwd_fail:
+	if (e_encoded_passwd) {
+		del_db2_cred(e_encoded_passwd);
+		del_db1_entry(e_encoded_passwd);
+		e_encoded_passwd = NULL;
+	}
+del_encoded_user_fail:
+del_db2_encoded_user_fail:
+	if (e_encoded_user) {
+		del_db2_cred(e_encoded_user);
+		del_db1_entry(e_encoded_user);
+		e_encoded_user = NULL;
+	}
 	/* Un-roll. */
 del_passwd_fail:
 	/* Fall-through expected */
@@ -1345,22 +1477,38 @@ del_db2_passwd_fail:
 		del_db1_entry(e_passwd);
 		e_passwd = NULL;
 	}
-
 del_user_fail:
+del_db2_user_fail:
 	if (e_user) {
 		del_db1_entry(e_user);
+		del_db2_cred(e_user);
 		e_user = NULL;
 	}
-
-del_db2_user_fail:
+encoded_passwd_dne:
+encoded_user_dne:
 	/* Fall-through expected */
 passwd_dne:
 	/* Fall-through expected */
 user_dne:
-
-	free(user);
+	if (encoded_user) {
+		free(encoded_user);
+		user = NULL;
+	}
+no_encoded_user_mem:
+	if (encoded_passwd) {
+		free(encoded_passwd);
+		encoded_passwd = NULL;
+	}
+no_encoded_passwd_mem:
+	if (user) {
+		free(user);
+		user = NULL;
+	}
 no_user_mem:
-	free(passwd);
+	if (passwd) {
+		free(passwd);
+		passwd = NULL;
+	}
 no_passwd_mem:
 	return ret;
 }
@@ -1384,41 +1532,70 @@ no_out_mem:
 }
 
 
-	/* Must add two entries (username/password) */
+/* At least add two entries (username/password), if encoded username/password
+ * is different with original one, another entry is added */
 int add(struct anspass_packet *in) {
 	int ret = -ENOMEM;
 	struct db1_entry *e_user = NULL;
 	struct db1_entry *e_passwd = NULL;
+
+	struct db1_entry *e_encoded_user = NULL;
+	struct db1_entry *e_encoded_passwd = NULL;
+
 	char* passwd = (char*)calloc(1, get_passwd_len(in->msg, in->user));
 	if (!passwd)
 		goto no_passwd_mem;
-
-
 	char* user = (char*)calloc(1, sizeof(char) * (strlen(in->msg) +
 				strlen(USERNAME_STR) + 5));
 	if (!user)
 		goto no_user_mem;
+
+	char* encoded_passwd = (char*)calloc(1, 3 * get_passwd_len(in->msg, in->user));
+	if (!encoded_passwd)
+		goto no_encoded_passwd_mem;
+	char* encoded_user = (char*)calloc(1, 3 * (sizeof(char) * (strlen(in->msg) +
+				strlen(USERNAME_STR) + 5)));
+	if (!encoded_user)
+		goto no_encoded_user_mem;
+
 	/* Username string is easy */
 	sprintf(user, "%s'%s': ", USERNAME_STR, in->msg);
 
+	get_percentencode_username_str(encoded_user, in);
+
 	/* Passwd has to insert username into the URI */
 	get_passwd_str(passwd, in);
 
+	get_percentencode_passwd_str(encoded_passwd, in);
 
 	ret = get_db1_entry(user, &e_user);
-	if (!ret)
-	{
+	if (!ret) {
 		ret = -EEXIST;
 		goto user_exists;
 	}
 
 	ret = get_db1_entry(passwd, &e_passwd);
-	if (!ret)
-	{
+	if (!ret) {
 		ret = -EEXIST;
 		goto passwd_exists;
 	}
 
+	if (strcmp(user, encoded_user) != 0) {
+		ret = get_db1_entry(encoded_user, &e_encoded_user);
+		if (!ret) {
+			ret = -EEXIST;
+			goto encoded_user_exists;
+		}
+	}
+
+	if (strcmp(passwd, encoded_passwd) != 0) {
+		ret = get_db1_entry(encoded_passwd, &e_encoded_passwd);
+		if (!ret) {
+			ret = -EEXIST;
+			goto encoded_passwd_exists;
+		}
+	}
+
 	ret = create_db1_entry(user, &e_user);
 	if (ret)
 		goto create_user_fail;
@@ -1435,11 +1612,35 @@ int add(struct anspass_packet *in) {
 	if (ret)
 		goto create_db2_passwd_fail;
 
+	if (strcmp(user, encoded_user) != 0) {
+		ret = create_db1_entry(encoded_user, &e_encoded_user);
+		if (ret)
+			goto create_encoded_user_fail;
 
+		ret = create_db2_entry(e_encoded_user, in->user);
+		if (ret)
+			goto create_db2_encoded_user_fail;
+	}
 
+	if (strcmp(passwd, encoded_passwd) != 0) {
+		ret = create_db1_entry(encoded_passwd, &e_encoded_passwd);
+		if (ret)
+			goto create_encoded_passwd_fail;
 
-	/* Un-roll & answer. */
-
+		ret = create_db2_entry(e_encoded_passwd, in->passwd);
+		if (ret)
+			goto create_db2_encoded_passwd_fail;
+	}
+create_db2_encoded_passwd_fail:
+	if (ret)
+		del_db1_entry(e_encoded_passwd);
+create_encoded_passwd_fail:
+	if (ret)
+		del_db2_cred(e_encoded_user);
+create_db2_encoded_user_fail:
+	if (ret)
+		del_db1_entry(e_encoded_user);
+create_encoded_user_fail:
 	if (ret)
 		del_db2_cred(e_passwd);
 create_db2_passwd_fail:
@@ -1452,16 +1653,35 @@ create_db2_user_fail:
 	if (ret)
 		del_db1_entry(e_user);
 create_user_fail:
+encoded_passwd_exists:
+encoded_user_exists:
 	/* Fall-through expected */
 passwd_exists:
 	/* Fall-through expected */
 user_exists:
-	free(user);
+	if (encoded_user) {
+		free(encoded_user);
+		encoded_user = NULL;
+	}
+no_encoded_user_mem:
+	if (encoded_passwd) {
+		free(encoded_passwd);
+		encoded_passwd = NULL;
+	}
+no_encoded_passwd_mem:
+	if (user) {
+		free(user);
+		user = NULL;
+	}
 no_user_mem:
-	free(passwd);
+	if (passwd) {
+		free(passwd);
+		passwd = NULL;
+	}
 no_passwd_mem:
 	return ret;
 }
+
 void process_add(struct anspass_packet *in) {
 	int ret = 0;
 
-- 
2.34.1

